### PR TITLE
Change page-table from a linear table to a tree

### PR DESCRIPTION
**Pull request by fischer-martin**
@Blazefrost and I changed the page-table to use a tree instead of a linear table. This reduces memory consumption in basically every case drastically.

### CODE REVIEW

**Comment by ckirsch** [comment]
How much slower does `make all` get? If it does get significantly slower, you probably need to measure by how much memory access slows down and, separately, by how much context switching slows down, to see if we need to do something about that.

File: selfie.c [hunk 1]
```diff
@@ -9443,15 +9506,36 @@ void map_page(uint64_t* context, uint64_t page, uint64_t frame) {
...
+  }
+
+  return translated_table;
+}
+
 void restore_region(uint64_t* context, uint64_t* table, uint64_t* parent_table, uint64_t lo, uint64_t hi) {
   uint64_t frame;
+  uint64_t* deref_table;
+
+  deref_table = dereference_root_table(parent_table, table);
```
Comment (ckirsch) [inline selfie.c:9533]: Is this really necessary? How about generalizing `is_virtual_address_mapped`?

**Comment by fischer-martin** [comment]
> How much slower does `make all` get? If it does get significantly slower, you probably need to measure by how much memory access slows down and, separately, by how much context switching slows down, to see if we need to do something about that.

comparison of `make all` runtimes:  
make_all

We'll definitely gonna measure the time differences for memory access and context switching between `master` and `tree-pt`.

**Comment by ckirsch** [comment]
Oops. So, that’s the issue really. I wonder if we need to go back to an array then but smaller up to the size of physical memory and then a separate stack page that can be accessed fast. Check if all stack accesses are done relative to the sp register. If yes, you may be able to modify the validity check for virtual addresses to make sure that nothing above the array-mapped memory is accessed other than through the sp register.

**Comment by Blazefrost** [comment]
> Check if all stack accesses are done relative to the sp register.

We looked into the compiler and stack accesses are done relative to REG_SP (for return address and call arguments)  or REG_FP (for local variables).

We also checked how many pages are required by `make all` and it only required a single page, save for `minster` which maps all pages at the beginning of execution.

> I wonder if we need to go back to an array then but smaller up to the size of physical memory and then a separate stack page that can be accessed fast.

We've considered this idea, but we fear that excluding a single stack page from the page table would introduce a lot of changes because it would require to separately handle this case in a lot of functions related to virtual memory. Furthermore, if additional stack pages are required, it would not be trivial to add these.

--------------------
In the recent changeset, we've introduced two optimizations to our PR:

* In commit `d59ccb7`, we've hard-coded the factors for the calculation of the first level index and second level index instead of invoking `left_shift` and `right_shift`. These changes decreased execution time of `make all` from 45.5 minutes to 37.25 minutes.
* In commit `7a28a4c`, we've added the cached virtual context root node into the context structure and perform updates to it only after new pages have been mapped. This further reduced execution time of `make all` down to 15 minutes, which is comparable to the current master with 14.25 minutes.

**Comment by Blazefrost** [comment]
We were successful in getting rid of the `dereference_root_table` function. In the end, the only function that required an additional reference to the hypervisor page table to get hypster running again was `page_for_frame`, which is used in `restore_region` only.

Changing `get_frame_for_page` was not necessary because hypster indirectly passes pages in its own address space to its contexts because `palloc` is page-aligned. This is exploited in `restore_region` where the frame for the virtual context's frame is mapped into the cached context (`selfie.c:9493`)

All changes to `restore_region` were reverted.

**Comment by Blazefrost** [comment]
We also adapted terminology from the RISC-V privileged specification, as described in commit 
22f8ca7.

Furthermore, the calculation for the root page table index and the leaf page table index are not hard-coded anymore, but are based on the count of leaf page table entries.

**Comment by fischer-martin** [comment]
There are two things we aren't sure about though:
* The name `hypervisor_table` for the new parameter in `frame_for_page` (https://github.com/cksystemsteaching/selfie/blob/16841e94c86b9fa279529009bc36472e4fb4c632/selfie.c#L6815). It's different from how that table is named normally (which is `parent_table`) but we thought it would be easier to understand this way.
* The size calculation for the root page-table (https://github.com/cksystemsteaching/selfie/blob/16841e94c86b9fa279529009bc36472e4fb4c632/selfie.c#L9200). We thought about ways to make this expression more readable for quite some time but sadly we couldn't find a good solution.

File: selfie.c [hunk 2]
```diff
@@ -1040,7 +1040,10 @@ void init_memory(uint64_t megabytes);
 uint64_t load_physical_memory(uint64_t* paddr);
 void     store_physical_memory(uint64_t* paddr, uint64_t data);
 
-uint64_t frame_for_page(uint64_t* table, uint64_t page);
+uint64_t root_page_table_index_for_page(uint64_t page);
```
Comment (ckirsch) [inline selfie.c:1043]: In fact, how about just `root_PTE`?

File: selfie.c [hunk 3]
```diff
@@ -1040,7 +1040,10 @@ void init_memory(uint64_t megabytes);
...
+uint64_t leaf_page_table_index_for_page(uint64_t page);
```
Comment (ckirsch) [inline selfie.c:1044]: And `leaf_PTE`?

File: selfie.c [hunk 4]
```diff
@@ -1040,7 +1040,10 @@ void init_memory(uint64_t megabytes);
...
+
+uint64_t frame_for_page(uint64_t* table, uint64_t* hypervisor_table, uint64_t page);
```
Comment (ckirsch) [inline selfie.c:1046]: I'd still prefer `parent_table` over `hypervisor_table`. The parent is the context that created the context of `table`...

File: selfie.c [hunk 5]
```diff
@@ -1067,6 +1070,8 @@ uint64_t WORDSIZEINBITS = 32;
 uint64_t INSTRUCTIONSIZE = 4; // must be the same as WORDSIZE
 uint64_t REGISTERSIZE    = 8; // must be twice of WORDSIZE
 
+uint64_t LEAF_PAGE_TABLE_ENTRIES = 512; // i.e. one leaf page table has the size PAGESIZE
```
Comment (ckirsch) [inline selfie.c:1073]: How about `uint64_t NUMBER_OF_LEAF_PTES = 512; // == PAGESIZE / REGISTERSIZE`?

File: selfie.c [hunk 6]
```diff
@@ -6799,12 +6804,35 @@ void store_physical_memory(uint64_t* paddr, uint64_t data) {
...
-  return (uint64_t) (table + page);
+uint64_t root_page_table_index_for_page(uint64_t page) {
+  return (page / LEAF_PAGE_TABLE_ENTRIES);
+}
+
+uint64_t leaf_page_table_index_for_page(uint64_t page) {
+  return (page - root_page_table_index_for_page(page) * LEAF_PAGE_TABLE_ENTRIES);
+}
+
+uint64_t frame_for_page(uint64_t* table, uint64_t* hypervisor_table, uint64_t page) {
```
Comment (ckirsch) [inline selfie.c:6815]: Please list `parent_table` as first parameter.

File: selfie.c [hunk 7]
```diff
@@ -6799,12 +6804,35 @@ void store_physical_memory(uint64_t* paddr, uint64_t data) {
...
 }
 
 uint64_t get_frame_for_page(uint64_t* table, uint64_t page) {
-  return *(table + page);
+  uint64_t* leaf_pt;
+
+  leaf_pt = (uint64_t*) *(table + root_page_table_index_for_page(page));
+
+  if (leaf_pt == (uint64_t*) 0)
+    // page isn't mapped if the corresponding internal node isn't mapped
```
Comment (ckirsch) [inline selfie.c:6832]: `// page is unmapped if its leaf PTE is unmapped`

File: selfie.c [hunk 8]
```diff
@@ -6799,12 +6804,35 @@ void store_physical_memory(uint64_t* paddr, uint64_t data) {
...
+
+  if (leaf_pt == (uint64_t*) 0)
+    return 0;
+  else
+    return (uint64_t) (leaf_pt + leaf_page_table_index_for_page(page));
 }
 
 uint64_t get_frame_for_page(uint64_t* table, uint64_t page) {
-  return *(table + page);
+  uint64_t* leaf_pt;
```
Comment (ckirsch) [inline selfie.c:6827]: `leaf_pte` rather than `leaf_pt`?

File: selfie.c [hunk 9]
```diff
@@ -9416,12 +9444,24 @@ void save_context(uint64_t* context) {
 
 void map_page(uint64_t* context, uint64_t page, uint64_t frame) {
   uint64_t* table;
+  uint64_t* leaf_pt;
+  uint64_t  root_pt_index;
```
Comment (ckirsch) [inline selfie.c:9448]: `root_pte`?

**Comment by fischer-martin** [comment]
> try to include code that allows going back to a flat page table

Do you mean a command-line argument that specifies which data structure is used?

**Comment by ckirsch** [comment]
> > try to include code that allows going back to a flat page table
> 
> Do you mean a command-line argument that specifies which data structure is used?

No. Just a flag in the source code will do.

File: selfie.c [hunk 10]
```diff
@@ -1065,8 +1068,15 @@ uint64_t WORDSIZEINBITS = 32;
 uint64_t INSTRUCTIONSIZE = 4; // must be the same as WORDSIZE
 uint64_t REGISTERSIZE    = 8; // must be twice of WORDSIZE
 
+uint64_t NUMBER_OF_LEAF_PTES = 512; // == PAGESIZE / REGISTERSIZE
+
 uint64_t PAGESIZE = 4096; // we use standard 4KB pages
 
+uint64_t PAGE_TABLE_LINEAR = 0; // use a linear page table
+uint64_t PAGE_TABLE_TREE   = 1; // use a two-level tree page table
```
Comment (ckirsch) [inline selfie.c:1076]: Just use PAGE_TABLE_TREE as flag.

**Pull request merged**